crates/fs/src/encodings.rs

```diff
@@ -0,0 +1,128 @@
+    sync::{Arc, Mutex},
+};
+
+use anyhow::Result;
+use encoding_rs::Encoding;
+
+/// A wrapper around `encoding_rs::Encoding` to implement `Send` and `Sync`.
+/// Since the reference is static, it is safe to send it across threads.
+pub struct EncodingWrapper(&'static Encoding);
+
+impl Debug for EncodingWrapper {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple(&format!("EncodingWrapper{:?}", self.0))
+            .field(&self.0.name())
+            .finish()
+    }
+}
+
+pub struct EncodingWrapperVisitor;
+
+impl PartialEq for EncodingWrapper {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.name() == other.0.name()
+    }
+}
+
+unsafe impl Send for EncodingWrapper {}
+unsafe impl Sync for EncodingWrapper {}
+
+impl Clone for EncodingWrapper {
+    fn clone(&self) -> Self {
+        EncodingWrapper(self.0)
+    }
+}
+
+impl EncodingWrapper {
+    pub fn new(encoding: &'static Encoding) -> EncodingWrapper {
+        EncodingWrapper(encoding)
+    }
+
+    pub fn get_encoding(&self) -> &'static Encoding {
+        self.0
+    }
+
+    pub async fn decode(
+        &mut self,
+        input: Vec<u8>,
+        force: bool,
+        buffer_encoding: Option<Arc<Mutex<&'static Encoding>>>,
+    ) -> Result<String> {
+        // Check if the input starts with a BOM for UTF-16 encodings only if not forced to
+        // use the encoding specified.
+        if !force {
+            if input.len() >= 2 {
+                if (input[0] == 0xFF) & (input[1] == 0xFE) {
+                    self.0 = encoding_rs::UTF_16LE;
+
+                    if let Some(v) = buffer_encoding {
+                        if let Ok(mut v) = (*v).lock() {
+                            *v = encoding_rs::UTF_16LE;
+                        }
+                    }
+                } else if (input.len() >= 2) & (input[0] == 0xFE) & (input[1] == 0xFF) {
+                    self.0 = encoding_rs::UTF_16BE;
+
+                    if let Some(v) = buffer_encoding {
+                        if let Ok(mut v) = (*v).lock() {
+                            *v = encoding_rs::UTF_16BE;
+                        }
+                    }
+                }
+            }
+        }
+
+        let (cow, _had_errors) = self.0.decode_with_bom_removal(&input);
+
+        // `encoding_rs` handles invalid bytes by replacing them with replacement characters
+        // in the output string, so we return the result even if there were errors.
+        // This preserves the original behaviour where files with invalid bytes could still be opened.
+        Ok(cow.into_owned())
+    }
+
+    pub async fn encode(&self, input: String) -> Result<Vec<u8>> {
+        if self.0 == encoding_rs::UTF_16BE {
+            let mut data = Vec::<u8>::with_capacity(input.len() * 2);
+
+            // Convert the input string to UTF-16BE bytes
+            let utf16be_bytes: Vec<u8> =
+                input.encode_utf16().flat_map(|u| u.to_be_bytes()).collect();
```

Comment:
`collect()` is unnecessary since it allocates and `extend` accepts iterator.
```suggestion
            let utf16be_bytes =
                input.encode_utf16().flat_map(|u| u.to_be_bytes());
```